datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.commons.lang3.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linkedin.datastream.common.PollUtils;
+import com.linkedin.datastream.common.RetriesExhaustedException;
+import com.linkedin.datastream.common.zk.ZkClient;
+import com.linkedin.datastream.server.ClusterThroughputInfo;
+import com.linkedin.datastream.server.DatastreamGroup;
+import com.linkedin.datastream.server.DatastreamGroupPartitionsMetadata;
+import com.linkedin.datastream.server.DatastreamSourceClusterResolver;
+import com.linkedin.datastream.server.DatastreamTask;
+import com.linkedin.datastream.server.providers.PartitionThroughputProvider;
+
+
+/**
+ * Partition assignment strategy that does assignment based on throughput information supplied by a
+ * {@link PartitionThroughputProvider}
+ */
+public class LoadBasedPartitionAssignmentStrategy extends StickyPartitionAssignmentStrategy {
+  private static final Logger LOG = LoggerFactory.getLogger(LoadBasedPartitionAssignmentStrategy.class.getName());
+
+  // TODO Make these constants configurable
+  private static final long THROUGHPUT_INFO_FETCH_TIMEOUT_MS_DEFAULT = Duration.ofSeconds(10).toMillis();
+  private static final long THROUGHPUT_INFO_FETCH_RETRY_PERIOD_MS_DEFAULT = Duration.ofSeconds(1).toMillis();
+
+  private final PartitionThroughputProvider _throughputProvider;
+  private final DatastreamSourceClusterResolver _sourceClusterResolver;
+
+
+  /**
+   * Creates an instance of {@link LoadBasedPartitionAssignmentStrategy}
+   */
+  public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughputProvider,
+      DatastreamSourceClusterResolver sourceClusterResolver, Optional<Integer> maxTasks,
+      Optional<Integer> imbalanceThreshold, Optional<Integer> maxPartitionPerTask, boolean enableElasticTaskAssignment,
+      Optional<Integer> partitionsPerTask, Optional<Integer> partitionFullnessFactorPct, Optional<ZkClient> zkClient,
+      String clusterName) {
+    super(maxTasks, imbalanceThreshold, maxPartitionPerTask, enableElasticTaskAssignment, partitionsPerTask,
+        partitionFullnessFactorPct, zkClient, clusterName);
+    _throughputProvider = throughputProvider;
+    _sourceClusterResolver = sourceClusterResolver;
+  }
+
+  @Override
+  public Map<String, Set<DatastreamTask>> assignPartitions(Map<String, Set<DatastreamTask>> currentAssignment,
+      DatastreamGroupPartitionsMetadata datastreamPartitions) {
+    DatastreamGroup datastreamGroup = datastreamPartitions.getDatastreamGroup();
+    String datastreamGroupName = datastreamGroup.getName();
+    Map<String, ClusterThroughputInfo> partitionThroughputInfo;
+
+    // Attempting to retrieve partition throughput info with a fallback mechanism to StickyPartitionAssignmentStrategy
+    try {
+      partitionThroughputInfo = fetchPartitionThroughputInfo();
```

Comment:
We need to be super careful on how often fetchPartitionThroughputInfo() is called. We might want to add code here to avoid calling this too often, skip this step, throw an alert and move on.

Reply:
We discussed this with @vmaheshw as well. We'll only be doing throughput based partition assignment initially (when no partitions are assigned), and not incrementally. I updated the PR with that change.